splice/queries/tile.py

```diff
@@ -1,45 +1,67 @@
-from splice.models import Adgroup, Tile
+from splice.models import Adgroup, AdgroupCategory, Tile
 from sqlalchemy.orm.exc import NoResultFound
 from sqlalchemy.exc import InvalidRequestError
-from sqlalchemy.sql import exists
+from sqlalchemy.sql import exists, func
 
 from splice.queries.common import row_to_dict
 
 
-def get_tiles_by_adgroup_id(adgroup_id):
-    from splice.environment import Environment
-
-    env = Environment.instance()
-
-    rows = (
-        env.db.session
-        .query(Tile)
-        .filter(Tile.adgroup_id == adgroup_id)
-        .order_by(Tile.id.desc())
-        .all()
-    )
-    output = [row_to_dict(d) for d in rows]
+def needsAdgroupJoin(filters):
+    adgroup_filters = ('adgroup_type', 'channel_id', 'locale')
+    return all(k in filters for k in adgroup_filters)
 
-    return output
 
+def get_tiles(campaign_id=None, adgroup_id=None, group_by=None,
+              limit_fields=None, filters=None):
```

Comment:
:+1: super powerful Swiss-Knife function. Takes time to walk through it though.

Perhaps we can refactor it a bit for the readability, like break it down to a few sub-functions, base query builder, then add join tables and group by columns, finally inject filters.